flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.hadoop.io.IntWritable;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+
+
+/**
+ * Util for any ObjectInspector related inspection and conversion of Hive data to/from Flink data.
+ *
+ * <p>Hive ObjectInspector is a group of flexible APIs to inspect value in different data representation,
+ * and developers can extend those API as needed, so technically, object inspector supports arbitrary data type in java.
+ */
+@Internal
+public class HiveInspectors {
+
+	/**
+	 * Get conversion for converting Flink object to Hive object from an ObjectInspector.
+	 */
+	public static HiveObjectConversion getConversion(ObjectInspector inspector) {
+		if (inspector instanceof PrimitiveObjectInspector) {
+			if (inspector instanceof JavaBooleanObjectInspector) {
+				if (((JavaBooleanObjectInspector) inspector).preferWritable()) {
+					return o -> new BooleanWritable((Boolean) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaStringObjectInspector) {
+				if (((StringObjectInspector) inspector).preferWritable()) {
+					return o -> new Text((String) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaByteObjectInspector) {
+				if (((JavaByteObjectInspector) inspector).preferWritable()) {
+					return o -> new ByteWritable((Byte) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaShortObjectInspector) {
+				if (((JavaShortObjectInspector) inspector).preferWritable()) {
+					return o -> new ShortWritable((Short) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaIntObjectInspector) {
+				if (((JavaIntObjectInspector) inspector).preferWritable()) {
+					return o -> new IntWritable((Integer) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaLongObjectInspector) {
+				if (((JavaLongObjectInspector) inspector).preferWritable()) {
+					return o -> new LongWritable((Long) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaFloatObjectInspector) {
+				if (((JavaFloatObjectInspector) inspector).preferWritable()) {
+					return o -> new FloatWritable((Float) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaDoubleObjectInspector) {
+				if (((JavaDoubleObjectInspector) inspector).preferWritable()) {
+					return o -> new DoubleWritable((Double) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			}
+		}
+
+		throw new FlinkHiveUDFException(
+			String.format("Flink doesn't support convert object conversion for %s yet", inspector));
+	}
+
+	/**
+	 * Unwraps an object from Hive object to Flink object with an ObjectInspector.
+	 */
+	public static Object unwrap(ObjectInspector inspector, Object data) {
```

Comment:
Why not merge `unwrap` and `HiveObjectConversion`?
I think `HiveObjectConversion` should provide two method: `toHiveObject` and `toFlinkObject`.
Just like blink `DataFormatConverters`.

Reply:
that's might be doable. I need to spend some time to take a look at it because, unlike DataFormatConverters,  there's a subtle difference between Flink -> Hive and Hive -> Flink: 

From Flink-> Hive, it will always be thru `JavaXxxObjectInspector`. But from Hive -> Flink, it can be thru either `WritableXxxObjectInspector` or `JavaXxxObjectInspector`

Reply:
Yeah, I get it.
But I recommend aligning `getConversion` and `unwrap`, calling them similar names, and returning classes (rather than direct conversions).
It doesn't seem to take much effort.

Reply:
Giving it a re-evaluation, I think it's better not to do it now as I can't see much value out of it. The main reason you choose the `HiveObjectConversion` way is to optimize conversions of mult input args, but it doesn't apply to output arg as it's always only one. The current structure of "unwrap" is the same as that in Spark, making it easy for us to compare and question on. 

I will rename it to "toFlinkObject(...)" to make the names more aligned

Reply:
> The current structure of "unwrap" is the same as that in Spark, making it easy for us to compare and question on.

In fact, Spark unwrap return a Function instead of runtime `instanceof`.